### DISCUSSION TITLE
Safari shows blue outline on a link after clicking and then pressing the back button [up-follow]

### DISCUSSION THREAD

**Discussion by tboerstad** [unpoly/unpoly#554] [Q&A] [answered]
How can I avoid the jarring blue outline when clicking the back button?

This is the minimal example I was able to come up with. 
See the video for details of what happens, namely that there's a blue outline around the clicked link when clicking the back button. This does not happen when not using unpoly (or HTMX which I migrated from).

BTW: Kudos for unpoly, impressive work. 

index.html & world.html (the two are identical)
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <script src="https://cdn.jsdelivr.net/npm/unpoly@3.5.0/unpoly.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/unpoly@3.5.0/unpoly.min.css" />
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    Hello, <a href="/world" up-follow>unpoly!</a>
  </body>
</html>
```

https://github.com/unpoly/unpoly/assets/4872288/a37ab31f-c1f4-4069-b759-ff30bc621a2c

**Comment by triskweline** [comment]
I think what happens is this:

- Your stylesheet has a style like `a:focus { outline: 3px solid deepskyblue; }`
- Clicking the link focused the link element
- Unpoly preserves focus-related state through history navigations. When you go back, the link is re-focused.

Browsers other than Safari lets you address this by only showing the outline for `:focus-visible` elements:

```html
a:focus:not(:focus-visible) { outline: none; }
```

Unfortunately Safari seems to use different heuristic for determining if an element with `:focus` is also `:focus-visible`.

**Reply by triskweline** [reply]
Some resources:

- [CSS Tricks: The :focus-visible Trick](https://css-tricks.com/the-focus-visible-trick/)
- [WICG: Should :focus-visible match when returning focus or programmatically focusing?](https://github.com/WICG/focus-visible/issues/88)
- [HTMLElement API: focus: options.focusVisible parameter](https://caniuse.com/mdn-api_htmlelement_focus_options_focusvisible_parameter)

**Reply by tboerstad** [reply]
Thanks for taking the time to answer @triskweline.
I think you're right about focus being the issue here

Two points:
1. I'm not using any stylesheet in the example, other than the one from unpoly. 
2. I get the same effect in Chrome, after adding`up-instant`

The payoff from adding `up-follow` and `up-instant` is a big win for unpoly.  It speed up my development tremendously .
The only downside I've seen is this jarring behaviour, but if it's only me who has noticed it either no one cares or I'm doing something wrong

Example for chrome:
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <script src="https://cdn.jsdelivr.net/npm/unpoly@3.5.0/unpoly.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/unpoly@3.5.0/unpoly.min.css" />
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    Hello, <a href="/world" up-follow up-instant>unpoly!</a>
  </body>
</html>
```


https://github.com/unpoly/unpoly/assets/4872288/66ad2bd6-f120-4f4f-97de-7bcfa7f759e6

**Reply by triskweline** [reply]
> I'm not using any stylesheet in the example, other than the one from unpoly.

You always get a default stylesheet from your browser.

> I get the same effect in Chrome, after adding `up-instant`

I can reproduce this. To be more exact, the link becomes `:focus-visible` if it is the first link I click on the page, but not if I click another link first.

I think it would be OK (for A11y) to hide the focus ring in this particular case. As `[up-instant]` works on `mousedown` we know that the user is using a pointing device and does not require a focus ring.

**Reply by triskweline** [reply, marked as answer]
Issue: [Hide focus ring when following an [up-instant] link](https://github.com/unpoly/unpoly/issues/555)

**Reply by tboerstad** [reply]
Awesome
Thanks for the detailed explanations too 🙏

**Comment by triskweline** [comment]
Fixed in 3.7.0.
